pytorch_lightning/trainer/data_loading.py

```diff
@@ -306,20 +308,17 @@ def _reset_eval_dataloader(
                 # percent or num_steps
                 limit_eval_batches = getattr(self, f'limit_{mode}_batches')
 
-                if num_batches != float('inf'):
-                    self._check_batch_limits(f'limit_{mode}_batches')
-
-                    # limit num batches either as a percent or num steps
-                    if isinstance(limit_eval_batches, float):
+                # limit num batches either as a percent or num steps
+                if isinstance(limit_eval_batches, int):
+                    num_batches = min(num_batches, limit_eval_batches)
+                else:
+                    if num_batches != float('inf'):
                         num_batches = int(num_batches * limit_eval_batches)
-                    else:
-                        num_batches = min(len(dataloader), limit_eval_batches)
-
-                elif limit_eval_batches not in (0.0, 1.0):
-                    raise MisconfigurationException(
-                        'When using an infinite DataLoader (e.g. with an IterableDataset'
-                        f' or when DataLoader does not implement `__len__`) for `limit_{mode}_batches`,'
-                        f' `Trainer(limit_{mode}_batches)` must be `0.0` or `1.0`.')
+                    elif limit_eval_batches not in (0.0, 1.0):
+                        raise MisconfigurationException(
+                            'When using an infinite DataLoader (e.g. with an IterableDataset'
+                            f' or when DataLoader does not implement `__len__`) for `limit_{mode}_batches`,'
+                            f' `Trainer(limit_{mode}_batches)` must be `0.0`, `1.0` or `int`')
```

Comment:
How can `limit_mode_batches = 0.0 or 1.0` for infinite dataloader?

Reply:
As far as I remember, a value of `0.0` should just run nothing and a value of `1.0` should run until the dataloader stops. If it is an actually infinite dataloader then it would just never end, so perhaps this should be written differently?

Reply:
yeah, as of now I just moved some if-else statements and with 0.0 it was neither working before nor it working now. I can change it as per requirement. But just like you said I agree it should work with both `0.0` and `int` and otherwise raise this exception with an updated statement here.

Reply:
and not just 1.0 but with every `0.0 < limit_mode_batches <= 1.0`

Reply:
@awaelchli what do you suggest here? I'll update and add tests accordingly.

Reply:
Hmm, we still need to support `limit_mode_batches = 1.0` for iterable datasets since you may want to just run until the iteration stops, without knowing the length of the dataloader ahead of time. I think the issue where this came up was #1252. I think this is what happens now as the enumerate call will just stop once the underlying iterator reaches a `StopIteration`.

Reply:
We could maybe have a warning when `limit_mode_batches = 1.0` to tell the user that this will go on until the iteration ends? (or just refer to them as something other than infinite dataloaders)

Reply:
We have the same for val_check_interval=1.0, where we distinguish between 1.0 (float, 100% of epoch) and 1 (int, batch count). I think it is reasonable to differentiate using the type and the docs make it very clear with the examples. Regarding this exception message, I think it should be changed to "When using an IterableDataset (e.g. of infinite size ...". Or just not refer to inf dataset as Ethan suggested.

Reply:
and for non-iterable datasets should 0.0 be supported with inf dataloader?

Reply:
and separate exceptions for Iterable and inf dataset? For Iterable `1.0` is supported and for inf dataloader `0.0` and `int` is supported?

Reply:
can you clarify. what would that look like? I don't know of any inf datasets that are not of type IterableDataset.

Reply:
Neither do I, never seen one. To he honest, I didn't even know about `IterableDataset` before this point 😅. Well I guess, I'll revert recent commits and change the exception message as suggested by Ethan.